File: app/src/main/java/it/liceoarzignano/bold/home/ShortcutHolder.java
package it.liceoarzignano.bold.home;

import android.content.Context;
import android.content.res.Resources;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import it.liceoarzignano.bold.R;
import it.liceoarzignano.bold.utils.UiUtils;

class ShortcutHolder extends RecyclerView.ViewHolder {
    private final CardView mCardView;
    private final ImageView mIcon;
    private final TextView mTitle;

    ShortcutHolder(View view) {
        super(view);
        mCardView = view.findViewById(R.id.home_item_shortcut_card);
        mIcon = view.findViewById(R.id.home_item_shortcut_icon);
        mTitle = view.findViewById(R.id.home_item_shortcut_title);
    }

    void bind(Context context, @StringRes int title, @DrawableRes int icon,
              @ColorRes int color, ShortcutListener listener) {
        int position = getAdapterPosition();
        mTitle.setText(context.getString(title));
        mIcon.setImageResource(icon);
        mCardView.setCardBackgroundColor(ContextCompat.getColor(context, color));
        mCardView.setOnClickListener(v -> listener.onClick(position));

        if (position != 0 && position != 4) {
            return;
        }

        // Fix margins to match the default grid
        Resources r = context.getResources();
        RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) mCardView.getLayoutParams();
        int defaultMargin = params.topMargin;
        int extraMargin = (int) UiUtils.dpToPx(r, 8);
        boolean isFirst = position == 0;
        params.setMargins(isFirst ? extraMargin : defaultMargin, defaultMargin,
                isFirst ? defaultMargin : extraMargin, defaultMargin);
    }

    interface ShortcutListener {
        void onClick(int position);
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/home/ShortcutHolder.kt
package it.liceoarzignano.bold.home

import android.content.Context
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import it.liceoarzignano.bold.R
import it.liceoarzignano.bold.utils.UiUtils

class ShortcutHolder(view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view) {
    private val mCardView: androidx.cardview.widget.CardView = view.findViewById(R.id.home_item_shortcut_card)
    private val mIcon: ImageView = view.findViewById(R.id.home_item_shortcut_icon)
    private val mTitle: TextView = view.findViewById(R.id.home_item_shortcut_title)

    fun bind(context: Context, @StringRes title: Int, @DrawableRes icon: Int,
             @ColorRes color: Int, listener: ShortcutListener) {
        val position = adapterPosition
        mTitle.text = context.getString(title)
        mIcon.setImageResource(icon)
        mCardView.setCardBackgroundColor(ContextCompat.getColor(context, color))
        mCardView.setOnClickListener { listener.onClick(position) }

        if (position != 0 && position != 4) {
            return
        }

        // Fix margins to match the default grid
        val r = context.resources
        val params = mCardView.layoutParams as androidx.recyclerview.widget.RecyclerView.LayoutParams
        val defaultMargin = params.topMargin
        val extraMargin = UiUtils.dpToPx(r, 16f).toInt()
        val isFirst = position == 0
        params.setMargins(if (isFirst) extraMargin else defaultMargin, defaultMargin,
                if (isFirst) defaultMargin else extraMargin, defaultMargin)
    }

    interface ShortcutListener {
        fun onClick(position: Int)
    }
}